src/com/ms/silverking/cloud/toporing/ProtoRegionList.java

```diff
@@ -17,368 +17,373 @@
     }
-    
-    double getOwnedFraction(Node owner) {
-        BigDecimal  totalOwned;
-        
-        totalOwned = BigDecimal.ZERO;
-        for (ProtoRegion region : protoRegions) {
-            if (region.contains(owner)) {
-                totalOwned = totalOwned.add(region.getRegion().getSizeBD(), LongRingspace.mathContext);
-            }
-        }
-        return totalOwned.divide(LongRingspace.bdSize, LongRingspace.mathContext).doubleValue();
+    return new ProtoRegionList(protoRegions);
+  }
+
+  /**
+   * Create a ProtoRegionList from the given nodes weighted according to the
+   * weightSpecs
+   *
+   * @param nodes
+   * @param weightSpecs
+   * @return
+   */
+  static ProtoRegionList create(List<Node> nodes, WeightSpecifications weightSpecs) {
+    List<ProtoRegion> protoRegions;
+    List<RingRegion> regions;
+
+    protoRegions = new ArrayList<>();
+    regions = RingRegion.allRingspace.divide(weightSpecs.getWeights(nodes));
+    for (RingRegion region : regions) {
+      protoRegions.add(new ProtoRegion(region, minPrimaryUnderFailure));
     }
-    
-    boolean isEmpty() {
-        return protoRegions.size() == 0;
+    return new ProtoRegionList(protoRegions);
+  }
+
+  //static ProtoRegionList create(Collection<Node> nodes, WeightSpecifications weightSpecs) {
+  //    return create(ImmutableList.copyOf(nodes), weightSpecs);
+  //}
+
+  /**
+   * Find next region that does not contain the given node
+   *
+   * @param protoRegions
+   * @param startIndex
+   * @param node
+   * @param replicaIndex
+   * @return
+   */
+  int nextRegion(int startIndex, Node node, int replicaIndex, ReplicationType rType) {
+    int index;
+    int regionsExamined;
+
+    if (debug) {
+      System.out.printf("nextRegion. startIndex %d node %s replicaIndex %d rType %s\n", startIndex, node.getIDString(),
+          replicaIndex, rType);
     }
-    
-    int size() {
-        return protoRegions.size();
+    assert startIndex >= 0;
+    assert replicaIndex >= 0;
+    assert node != null;
+    assert rType != null;
+    regionsExamined = 0;
+    index = startIndex;
+    while (regionsExamined < protoRegions.size() && (protoRegions.get(index).contains(node) || protoRegions.get(
+        index).totalOwners(rType) > replicaIndex)) {
+      if (debug) {
+        System.out.println(index + "\t" + protoRegions.get(index));
+      }
+      index = (index + 1) % protoRegions.size();
+      regionsExamined++;
     }
-    
-    long getFirstRegionStart() {
-        return protoRegions.get(0).getRegion().getStart();
+    if (!protoRegions.get(index).contains(node) && protoRegions.get(index).totalOwners(rType) <= replicaIndex) {
+      if (debug) {
+        System.out.printf("nextRegion. index %d\n", index);
+      }
+      return index;
+    } else {
+      if (debug) {
+        System.out.printf("nextRegion. Couldn't find region without node: %s\n", node.getIDString());
+      }
+      return -1;
     }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 171:181, 214:224.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.